rfcs/20190814-kernel-and-op-registration.md

```diff
@@ -0,0 +1,658 @@
+
+There are a few approaches to running kernels on GPUs:
+
+* Assign computation to Eigen device (for e.g. see `OneHot`, `Transpose`,
+  training ops). (>200 occurrences in TensorFlow)
+
+* Call `device.parallelFor` (for e.g. see `BatchSelect`). (4 occurrences)
+
+* Call `ThreadPool::ParallelFor` (for e.g. see `MatrixDiag`). This is a
+  TensorFlow wrapper that eventually wraps calls to Eigen. For example,
+  `ThreadPool::ParallelFor` calls `device.parallelFor` in Eigen. (29
+  occurrences)
+
+* Call `Shard` (e.g. `CTCGreedyDecoder`). This approach is deprecated in favor
+  of `ThreadPool::TransformRangeConcurrently` but no kernels use the latter yet.
+  (42 occurrences)
+
+* Call `GpuLaunchKernel` or `CudaLaunchKernel` directly, i.e. without calling Eigen.
+(58 occurrences)
+
+* `Matmul` op calls directly to `StreamExecutor`.
+
+* Possibly others
+
+In all approaches above, TensorFlow core is responsible for maintaining
+respective device queues, streams or pools. Kernels then use these queues to
+schedule computation. Therefore, our primary goal is to implement a C API that
+enables this scheduling. To give an example, one approach we can take is have
+Kernel pass a callback across C API. Tensorflow core would then call this
+callback. See diagram below:
+
+![device API](20190814-kernel-and-op-registration/device_api_overview.png)
+
+Furthermore, note that most of the approaches listed above eventually call to
+Eigen to parallelize and forward computation to device. For example, the first
+approach above uses Eigen APIs directly. Consequently, we need to understand how
+Eigen works with devices and in some cases make changes to Eigen codebase as
+well.
+
+Finally, we should aim to create a smaller API. Some of the approaches listed in
+the Background section seem to be very similar. For example, calling
+`parallelFor` in Eigen is quite similar to calling into
+`ThreadPool::ParallelFor`. Therefore, we will only provide C API equivalents for
+the following:
+
+* `ThreadPool` and its methods.
+
+* `CudaLaunchKernel` function.
+
+* Computation assignment to device in Eigen.
+
+This proposal focuses on these 3 components for now. Due to the complexity and
+variety of TensorFlow kernels, it is very likely that we will need to consider
+more approaches going forward. For example, how `MatMul` op would call
+`StreamExecutor` directly has not been investigated.
+
+### ThreadPool API
+
+Here, we can just wrap relevant methods in the `ThreadPool` class.
+
+```c++
+TF_CAPI_EXPORT extern void TF_ThreadPool_Schedule(
+    TF_OpKernelContext* context,
+    void (*fn)());
+
+TF_CAPI_EXPORT extern void TF_ThreadPool_ScheduleWithHint(
+    TF_OpKernelContext* context,
+    void (*fn)(),
+    int start,
+    int limit);
+
+TF_CAPI_EXPORT extern void TF_ThreadPool_ParallelFor(
+    TF_OpKernelContext* context,
+    int64_t total,
+    int64_t cost_per_unit,
+    void (*fn)(int64_t, int64_t));
+
+TF_CAPI_EXPORT extern void TF_ThreadPool_ParallelForWithWorkerId(
```

Comment:
One thing which is not obvious when reading this proposal is that in C++ we need to wrap almost every call to a method in the context with the op_requires_ok macro to terminate execution in case of an error.

The same will need to be done here as all of these can fail, so I think we need to be explicit about the code the user is expected to write.

Reply:
Context is just used to call accessors, for e.g.:
`auto thread_pool = context->device()->tensorflow_cpu_worker_threads()->workers;`

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/matrix_diag_op.cc#L323

We could modify ThreadPool interface though to also return a Status. Then, we can also take TF_Status output parameter in C API.

Reply:
I added a section below called "Getting Status when using device APIs". Basically, since the status is not on the surface of APIs that we are wrapping (we could add it to ThreadPool, but probably not to Eigen::StreamInterface), I propose to have a separate function call to get status.

Let me know if this sounds good.